### DISCUSSION TITLE
Unity Support?

### DISCUSSION THREAD

**Discussion by yottayuan** [AvaloniaUI/Avalonia#6364] [Q&A]
The WebSite says Avalonia could work with Unity ? How? Any Document?

**Comment by kekekeks** [comment]
Avalonia is _used by_ Unity in their version control system.

**Reply by yottayuan** [reply]
I really hope that one day I can use Avalonia in Unity。

**Reply by kekekeks** [reply]
It world help to prioritize that feature if you could describe your intended usage scenarios.

**Reply by yottayuan** [reply]
At present, we use Unity's own GUI-UGUI to write UI programs. If we can use Avalonia to write UI programs in Unity, it would be great. I know that NoesisGUI (https://www.noesisengine.com/) can achieve this goal, but they are closed source, and their license is too harsh for individual users.

**Reply by yottayuan** [reply]
https://github.com/TheKeyblader/Unilonia Is another case, but the author said, it''s not stable and be paused.

**Comment by Akeemovic** [comment]
This would be really great. I came here looking for NoesisGUI alternative as it is not indie friendly.

**Comment by aliir960** [comment]
please make it true. i really need xaml for front and UI in Unity. for now, i am using NoesisGUI, but it is close source and also its license is expensive. we use Unity for creating industrial and educational applications and these applications need Complex UI and also Vector base UI which Unity haven't them(there is UI Tookit but it is not mature). i am looking for such capability in Unity for many years...

**Reply by thevortexcloud** [reply]
Do you need it in Unity or with Unity? If you need it with Unity you can do this pretty easily already by just hosting Unity inside a native control and then drawing the Avalonia content around it. 

The Mono game implementation is probably also a decent reference for this sort of thing:

https://github.com/AvaloniaInside/AvaloniaInside.MonoGame

You can then communicate with Unity via some sort of IPC.

If you need it in Unity then it gets a bit more complicated. If Unity allows for rendering to a texture it _should_ be possible. The Godot implementation meanwhile is probably a good example of that as it provides a full Avalonia platform implementation for Godot:

https://github.com/MrJul/Estragonia

Otherwise I imagine you could probably pay the Avalonia devs to get official Unity support in.

**Reply by najak3d** [reply]
And what would that "cost"? 

Call it: Unilonia.

I'm in Unity now, and NoesisGUI seems to be the only decent option currently, for UI inside Unity.

Avalonia would be better.

The Use case is for "games".  Hosting Unity in an app and putting Avalonia on top is not a good solution - it's too "separated" without a great way to edit/restart/debug Unity as you do development. 

We need a UI (i.e. Avalonia) inside of Unity itself, where the C# of the UI itself can also be debugged, inside the same IDE as the rest of the Unity scripts.

I'm 100% for pay-the-developers.   Does Avalonia have a way to do this now?  (group funding)

**Reply by thevortexcloud** [reply]
> And what would that "cost"?

No idea. You would probably have to get a quote. I suspect it will not be cheap though as it's a whole new backend.

> I'm 100% for pay-the-developers. Does Avalonia have a way to do this now? (group funding)

They do indeed. But it's not really meant for crowd sourcing. It's more company contract application work focused. You can schedule a job request here:

https://avaloniaui.net/services

I imagine though they don't really care where the money comes from as long as the they get it though. @MikeCodesDotNET (sorry for the ping) can probably give some general indication/steps required for this.

**Comment by theoractice** [comment]
If you need App-like UI in Unity, maybe you can try [UIWidgets](https://github.com/UIWidgets/com.unity.uiwidgets).
Basically it's like using C# to develop flutter applications. I'd say this works strangely well on PC.
![image](https://camo.githubusercontent.com/3e33a23b7dbb524b5893b7575b3c765a42ca1140dbb6ced4275e39157da36faa/68747470733a2f2f636f6e6e6563742d7072642d63646e2e756e6974792e636f6d2f32303139303332332f702f696d616765732f32613237363036662d613263632d346339662d396533342d6262333961653634643036635f756977696467657473312e676966)
Downside is that you must use the **Chinese** version of Unity.

**Reply by najak3d** [reply]
"UIWidgets" -- I'm not sure why you'd recommend it.  It looks like it was left undone, and abandoned about 3-4 years ago.    I'm talking about the github repo for the link you posted.

NoesisGUI seems to be the most promising UI for Unity that I've seen, although I haven't tried it yet.   I've tried several others, but to no good avail.  I was using NovaUI, 1.5 yrs ago, but they abandoned it before it was complete.   It seems that NoesisGUI has a solid business plan, and a long history.

But I'd prefer to see Avalonia insert itself into the Unity space here.

**Reply by theoractice** [reply]
> "UIWidgets" -- I'm not sure why you'd recommend it. It looks like it was left undone, and abandoned about 3-4 years ago. 

Unity China uses this GUI framework to develop their official Chinese Unity community app. The app itself is updating every month so they have no reason to abandon it as for now. I don't know why the github repo isn't updating, maybe it's feature complete or they just aren't ready to release the latest version. You surely cannot customize everything with this, but if someone want to try App-like UI in Unity then this is the fastest way.

**Reply by najak3d** [reply]
Well, what they have on Github appears to be for the older version of Unity (2020?) -- not Unity 6, and it's not getting updates.    And it's supposedly for "Chinese versions" - so not sure what that might mean for "English" support, as we'll be 2nd-class concern.

We're looking for something that can be considered "production quality" and well-supported.  I'd like to see Avalonia be more commercialized (to a point) for this cause - so that the maintainers are well-paid, as they should be.

NoesisGUI is paid - and this is good, but is a black box solution.   So when things go wrong - ....

With Avalonia, it's all C#, and open source, so our hands aren't tied if something goes wrong.  In which case we only need to make sure it's a "main stream" solution -- which it seems to be ahead of the competitors, so looks good here too.

**Reply by theoractice** [reply]
> We're looking for something that can be considered "production quality" and well-supported. I'd like to see Avalonia be more commercialized (to a point) for this cause - so that the maintainers are well-paid, as they should be.

Yes, I'd love Avalonia to appear on Unity too. 
UIWidgets is limited to Chinese users, and this situation is somehow related to the trade war between china and usa, so...

**Reply by najak3d** [reply]
> > We're looking for something that can be considered "production quality" and well-supported. I'd like to see Avalonia be more commercialized (to a point) for this cause - so that the maintainers are well-paid, as they should be.
> 
> Yes, I'd love Avalonia to appear on Unity too. UIWidgets is limited to Chinese users, and this situation is somehow related to the trade war between china and usa, so...

Tell them to just put a tariff on the Open Source cost -- and charge us double what they charge everyone else. :)

Joking aside, that sounds like a jibberish excuse, since the Github repo was abandoned 3 years ago, well before tariffs.
